File: PythonAPI/carla/agents/navigation/behavior_agent.py
# Copyright (c) # Copyright (c) 2018-2020 CVC.
#
# This work is licensed under the terms of the MIT license.
# For a copy, see <https://opensource.org/licenses/MIT>.


""" This module implements an agent that roams around a track following random
waypoints and avoiding other vehicles. The agent also responds to traffic lights,
traffic signs, and has different possible configurations. """

import random
import numpy as np
import carla
from agents.navigation.basic_agent import BasicAgent
from agents.navigation.local_planner import RoadOption
from agents.navigation.behavior_types import Cautious, Aggressive, Normal

from agents.tools.misc import get_speed, positive, is_within_distance, compute_distance

class BehaviorAgent(BasicAgent):
    """
    BehaviorAgent implements an agent that navigates scenes to reach a given
    target destination, by computing the shortest possible path to it.
    This agent can correctly follow traffic signs, speed limitations,
    traffic lights, while also taking into account nearby vehicles. Lane changing
    decisions can be taken by analyzing the surrounding environment such as tailgating avoidance.
    Adding to these are possible behaviors, the agent can also keep safety distance
    from a car in front of it by tracking the instantaneous time to collision
    and keeping it in a certain range. Finally, different sets of behaviors
    are encoded in the agent, from cautious to a more aggressive ones.
    """

    def __init__(self, vehicle, behavior='normal', opt_dict={}, map_inst=None, grp_inst=None):
        """
        Constructor method.

            :param vehicle: actor to apply to local planner logic onto
            :param behavior: type of agent to apply
        """

        super().__init__(vehicle, opt_dict=opt_dict, map_inst=map_inst, grp_inst=grp_inst)
        self._look_ahead_steps = 0

        # Vehicle information
        self._speed = 0
        self._speed_limit = 0
        self._direction = None
        self._incoming_direction = None
        self._incoming_waypoint = None
        self._min_speed = 5
        self._behavior = None
        self._sampling_resolution = 4.5

        # Parameters for agent behavior
        if behavior == 'cautious':
            self._behavior = Cautious()

        elif behavior == 'normal':
            self._behavior = Normal()

        elif behavior == 'aggressive':
            self._behavior = Aggressive()

    def _update_information(self):
        """
        This method updates the information regarding the ego
        vehicle based on the surrounding world.
        """
        self._speed = get_speed(self._vehicle)
        self._speed_limit = self._vehicle.get_speed_limit()
        self._local_planner.set_speed(self._speed_limit)
        self._direction = self._local_planner.target_road_option
        if self._direction is None:
            self._direction = RoadOption.LANEFOLLOW

        self._look_ahead_steps = int((self._speed_limit) / 10)

        self._incoming_waypoint, self._incoming_direction = self._local_planner.get_incoming_waypoint_and_direction(
            steps=self._look_ahead_steps)
        if self._incoming_direction is None:
            self._incoming_direction = RoadOption.LANEFOLLOW

    def traffic_light_manager(self):
        """
        This method is in charge of behaviors for red lights.
        """
        actor_list = self._world.get_actors()
        lights_list = actor_list.filter("*traffic_light*")
        affected, _ = self._affected_by_traffic_light(lights_list)

        return affected

    def _tailgating(self, waypoint, vehicle_list):
        """
        This method is in charge of tailgating behaviors.

            :param location: current location of the agent
            :param waypoint: current waypoint of the agent
            :param vehicle_list: list of all the nearby vehicles
        """

        left_turn = waypoint.left_lane_marking.lane_change
        right_turn = waypoint.right_lane_marking.lane_change

        left_wpt = waypoint.get_left_lane()
        right_wpt = waypoint.get_right_lane()

        behind_vehicle_state, behind_vehicle, _ = self._vehicle_obstacle_detected(vehicle_list, max(
            self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=180, low_angle_th=160)
        if behind_vehicle_state and self._speed < get_speed(behind_vehicle):
            if (right_turn == carla.LaneChange.Right or right_turn ==
                    carla.LaneChange.Both) and waypoint.lane_id * right_wpt.lane_id > 0 and right_wpt.lane_type == carla.LaneType.Driving:
                new_vehicle_state, _, _ = self._vehicle_obstacle_detected(vehicle_list, max(
                    self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=180, lane_offset=1)
                if not new_vehicle_state:
                    print("Tailgating, moving to the right!")
                    end_waypoint = self._local_planner.target_waypoint
                    self._behavior.tailgate_counter = 200
                    self.set_destination(end_waypoint.transform.location,
                                         right_wpt.transform.location)
            elif left_turn == carla.LaneChange.Left and waypoint.lane_id * left_wpt.lane_id > 0 and left_wpt.lane_type == carla.LaneType.Driving:
                new_vehicle_state, _, _ = self._vehicle_obstacle_detected(vehicle_list, max(
                    self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=180, lane_offset=-1)
                if not new_vehicle_state:
                    print("Tailgating, moving to the left!")
                    end_waypoint = self._local_planner.target_waypoint
                    self._behavior.tailgate_counter = 200
                    self.set_destination(end_waypoint.transform.location,
                                         left_wpt.transform.location)

    def collision_and_car_avoid_manager(self, waypoint):
        """
        This module is in charge of warning in case of a collision
        and managing possible tailgating chances.

            :param location: current location of the agent
            :param waypoint: current waypoint of the agent
            :return vehicle_state: True if there is a vehicle nearby, False if not
            :return vehicle: nearby vehicle
            :return distance: distance to nearby vehicle
        """

        vehicle_list = self._world.get_actors().filter("*vehicle*")
        def dist(v): return v.get_location().distance(waypoint.transform.location)
        vehicle_list = [v for v in vehicle_list if dist(v) < 45 and v.id != self._vehicle.id]

        if self._direction == RoadOption.CHANGELANELEFT:
            vehicle_state, vehicle, distance = self._vehicle_obstacle_detected(
                vehicle_list, max(
                    self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=180, lane_offset=-1)
        elif self._direction == RoadOption.CHANGELANERIGHT:
            vehicle_state, vehicle, distance = self._vehicle_obstacle_detected(
                vehicle_list, max(
                    self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=180, lane_offset=1)
        else:
            vehicle_state, vehicle, distance = self._vehicle_obstacle_detected(
                vehicle_list, max(
                    self._behavior.min_proximity_threshold, self._speed_limit / 3), up_angle_th=30)

            # Check for tailgating
            if not vehicle_state and self._direction == RoadOption.LANEFOLLOW \
                    and not waypoint.is_junction and self._speed > 10 \
                    and self._behavior.tailgate_counter == 0:
                self._tailgating(waypoint, vehicle_list)

        return vehicle_state, vehicle, distance

    def pedestrian_avoid_manager(self, waypoint):
        """
        This module is in charge of warning in case of a collision
        with any pedestrian.

            :param location: current location of the agent
            :param waypoint: current waypoint of the agent
            :return vehicle_state: True if there is a walker nearby, False if not
            :return vehicle: nearby walker
            :return distance: distance to nearby walker
        """

        walker_list = self._world.get_actors().filter("*walker.pedestrian*")
        def dist(w): return w.get_location().distance(waypoint.transform.location)
        walker_list = [w for w in walker_list if dist(w) < 10]

        if self._direction == RoadOption.CHANGELANELEFT:
            walker_state, walker, distance = self._vehicle_obstacle_detected(walker_list, max(
                self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=90, lane_offset=-1)
        elif self._direction == RoadOption.CHANGELANERIGHT:
            walker_state, walker, distance = self._vehicle_obstacle_detected(walker_list, max(
                self._behavior.min_proximity_threshold, self._speed_limit / 2), up_angle_th=90, lane_offset=1)
        else:
            walker_state, walker, distance = self._vehicle_obstacle_detected(walker_list, max(
                self._behavior.min_proximity_threshold, self._speed_limit / 3), up_angle_th=60)

        return walker_state, walker, distance

    def car_following_manager(self, vehicle, distance, debug=False):
        """
        Module in charge of car-following behaviors when there's
        someone in front of us.

            :param vehicle: car to follow
            :param distance: distance from vehicle
            :param debug: boolean for debugging
            :return control: carla.VehicleControl
        """

        vehicle_speed = get_speed(vehicle)
        delta_v = max(1, (self._speed - vehicle_speed) / 3.6)
        ttc = distance / delta_v if delta_v != 0 else distance / np.nextafter(0., 1.)

        # Under safety time distance, slow down.
        if self._behavior.safety_time > ttc > 0.0:
            target_speed = min([
                positive(vehicle_speed - self._behavior.speed_decrease),
                self._behavior.max_speed,
                self._speed_limit - self._behavior.speed_lim_dist])
            self._local_planner.set_speed(target_speed)
            control = self._local_planner.run_step(debug=debug)

        # Actual safety distance area, try to follow the speed of the vehicle in front.
        elif 2 * self._behavior.safety_time > ttc >= self._behavior.safety_time:
            target_speed = min([
                max(self._min_speed, vehicle_speed),
                self._behavior.max_speed,
                self._speed_limit - self._behavior.speed_lim_dist])
            self._local_planner.set_speed(target_speed)
            control = self._local_planner.run_step(debug=debug)

        # Normal behavior.
        else:
            target_speed = min([
                self._behavior.max_speed,
                self._speed_limit - self._behavior.speed_lim_dist])
            self._local_planner.set_speed(target_speed)
            control = self._local_planner.run_step(debug=debug)

        return control

    def run_step(self, debug=False):
        """
        Execute one step of navigation.

            :param debug: boolean for debugging
            :return control: carla.VehicleControl
        """
        self._update_information()

        control = None
        if self._behavior.tailgate_counter > 0:
            self._behavior.tailgate_counter -= 1

        ego_vehicle_loc = self._vehicle.get_location()
        ego_vehicle_wp = self._map.get_waypoint(ego_vehicle_loc)

        # 1: Red lights and stops behavior
        if self.traffic_light_manager():
            return self.emergency_stop()

        # 2.1: Pedestrian avoidance behaviors
        walker_state, walker, w_distance = self.pedestrian_avoid_manager(ego_vehicle_wp)

        if walker_state:
            # Distance is computed from the center of the two cars,
            # we use bounding boxes to calculate the actual distance
            distance = w_distance - max(
                walker.bounding_box.extent.y, walker.bounding_box.extent.x) - max(
                    self._vehicle.bounding_box.extent.y, self._vehicle.bounding_box.extent.x)

            # Emergency brake if the car is very close.
            if distance < self._behavior.braking_distance:
                return self.emergency_stop()

        # 2.2: Car following behaviors
        vehicle_state, vehicle, distance = self.collision_and_car_avoid_manager(ego_vehicle_wp)

        if vehicle_state:
            # Distance is computed from the center of the two cars,
            # we use bounding boxes to calculate the actual distance
            distance = distance - max(
                vehicle.bounding_box.extent.y, vehicle.bounding_box.extent.x) - max(
                    self._vehicle.bounding_box.extent.y, self._vehicle.bounding_box.extent.x)

            # Emergency brake if the car is very close.
            if distance < self._behavior.braking_distance:
                return self.emergency_stop()
            else:
                control = self.car_following_manager(vehicle, distance)

        # 3: Intersection behavior
        elif self._incoming_waypoint.is_junction and (self._incoming_direction in [RoadOption.LEFT, RoadOption.RIGHT]):
            target_speed = min([
                self._behavior.max_speed,
                self._speed_limit - 5])
            self._local_planner.set_speed(target_speed)
            control = self._local_planner.run_step(debug=debug)

        # 4: Normal behavior
        else:
            target_speed = min([
                self._behavior.max_speed,
                self._speed_limit - self._behavior.speed_lim_dist])
            self._local_planner.set_speed(target_speed)
            control = self._local_planner.run_step(debug=debug)

        return control

    def emergency_stop(self):
        """
        Overwrites the throttle a brake values of a control to perform an emergency stop.
        The steering is kept the same to avoid going out of the lane when stopping during turns

            :param speed (carl.VehicleControl): control to be modified
        """
        control = carla.VehicleControl()
        control.throttle = 0.0
        control.brake = self._max_brake
        control.hand_brake = False
        return control

File: LibCarla/source/carla/agents/navigation/BehaviorAgent.cpp
// Copyright (c) 2018-2020 CVC.
//
// This work is licensed under the terms of the MIT license.
// For a copy, see <https://opensource.org/licenses/MIT>.

#include "carla/agents/navigation/BehaviorAgent.h"

#include "carla/agents/navigation/LocalPlanner.h"
#include "carla/agents/navigation/Misc.h"
#include "carla/client/Actor.h"
#include "carla/client/ActorList.h"
#include "carla/client/Map.h"
#include "carla/client/Vehicle.h"
#include "carla/client/Waypoint.h"
#include "carla/client/World.h"
#include "carla/geom/BoundingBox.h"
#include "carla/geom/Location.h"
#include "carla/geom/Transform.h"
#include "carla/road/Lane.h"
#include "carla/road/element/LaneMarking.h"
#include "carla/rpc/VehicleControl.h"

#include <algorithm>
#include <cmath>
#include <iostream>
#include <type_traits>

namespace carla {
namespace agents {
namespace navigation {

namespace {

  using LaneChange = road::element::LaneMarking::LaneChange;

  /// True if a `LaneChange` value contains the `flag` bit
  /// (mirrors Python `==` checks like `right_turn == LaneChange.Right` /
  /// `right_turn == LaneChange.Both`).
  bool HasLaneChange(LaneChange value, LaneChange flag) {
    using U = std::underlying_type_t<LaneChange>;
    return (static_cast<U>(value) & static_cast<U>(flag)) == static_cast<U>(flag);
  }

  BehaviorParameters BehaviorParamsFor(BehaviorAgent::Behavior b) {
    switch (b) {
      case BehaviorAgent::Behavior::Cautious:   return BehaviorParameters::Cautious();
      case BehaviorAgent::Behavior::Aggressive: return BehaviorParameters::Aggressive();
      case BehaviorAgent::Behavior::Normal:
      default:                                  return BehaviorParameters::Normal();
    }
  }

} // anonymous namespace

// ---------------------------------------------------------------------------
// Construction
// ---------------------------------------------------------------------------

BehaviorAgent::BehaviorAgent(SharedPtr<client::Vehicle> vehicle,
                             Behavior behavior,
                             const Options &opts,
                             SharedPtr<client::Map> map,
                             SharedPtr<GlobalRoutePlanner> grp)
  : BasicAgent(vehicle, /*target_speed_kmh=*/20.0f, opts, map, grp),
    _behavior(BehaviorParamsFor(behavior)) {
  // Match the Python BehaviorAgent override of the sampling resolution.
  _sampling_resolution = 4.5f;
}

// ---------------------------------------------------------------------------
// Information refresh
// ---------------------------------------------------------------------------

void BehaviorAgent::UpdateInformation() {
  _speed = GetSpeed(_vehicle);
  _speed_limit = _vehicle->GetSpeedLimit();
  // Even when follow_speed_limits is on this seed mirrors the Python reference.
  _local_planner->SetSpeed(_speed_limit);

  _direction = _local_planner->TargetRoadOption();
  if (_direction == RoadOption::Void) {
    _direction = RoadOption::LaneFollow;
  }

  _look_ahead_steps = static_cast<int>(_speed_limit / 10.0f);

  auto incoming = _local_planner->GetIncomingWaypointAndDirection(_look_ahead_steps);
  _incoming_waypoint = incoming.first;
  _incoming_direction = incoming.second;
  if (_incoming_direction == RoadOption::Void) {
    _incoming_direction = RoadOption::LaneFollow;
  }
}

// ---------------------------------------------------------------------------
// Traffic light handling
// ---------------------------------------------------------------------------

bool BehaviorAgent::TrafficLightManager() {
  auto actor_list = _world->GetActors();
  auto lights_list = actor_list ? actor_list->Filter("*traffic_light*") : nullptr;
  auto result = AffectedByTrafficLight(lights_list);
  return result.first;
}

// ---------------------------------------------------------------------------
// Tailgating
// ---------------------------------------------------------------------------

void BehaviorAgent::Tailgating(SharedPtr<client::Waypoint> waypoint) {
  // Python passes the prefiltered vehicle list around; the base helper
  // `VehicleObstacleDetected` accepts a `SharedPtr<ActorList>` and applies
  // its own proximity gating via `max_distance`, so we hand it the
  // world-wide vehicle filter result here.
  auto left_marking = waypoint->GetLeftLaneMarking();
  auto right_marking = waypoint->GetRightLaneMarking();
  auto left_wpt = waypoint->GetLeft();
  auto right_wpt = waypoint->GetRight();

  // Build the world-vehicle ActorList passthrough (matches Python's reuse of
  // the filtered list).
  auto vehicles_alist = _world->GetActors()
      ? _world->GetActors()->Filter("*vehicle*")
      : nullptr;

  const float behind_max =
      std::max(_behavior.min_proximity_threshold, _speed_limit / 2.0f);

  auto behind = VehicleObstacleDetected(vehicles_alist,
                                        behind_max,
                                        /*up_angle_th=*/180.0f,
                                        /*low_angle_th=*/160.0f,
                                        /*lane_offset=*/0);

  if (!(behind.affected && behind.actor)) return;
  if (_speed >= GetSpeed(behind.actor)) return;

  const int wp_lane_id = waypoint->GetLaneId();

  // Right-lane tailgate avoidance.
  if (right_marking.has_value() && right_wpt &&
      (HasLaneChange(right_marking->lane_change, LaneChange::Right) ||
       HasLaneChange(right_marking->lane_change, LaneChange::Both)) &&
      (wp_lane_id * right_wpt->GetLaneId() > 0) &&
      right_wpt->GetType() == road::Lane::LaneType::Driving) {
    auto check = VehicleObstacleDetected(vehicles_alist,
                                         behind_max,
                                         /*up_angle_th=*/180.0f,
                                         /*low_angle_th=*/0.0f,
                                         /*lane_offset=*/1);
    if (!check.affected) {
      std::cout << "Tailgating, moving to the right!" << std::endl;
      auto end_waypoint = _local_planner->TargetWaypoint();
      _behavior.tailgate_counter = 200;
      if (end_waypoint) {
        SetDestination(end_waypoint->GetTransform().location,
                       right_wpt->GetTransform().location);
      }
    }
    return;
  }

  // Left-lane tailgate avoidance.
  if (left_marking.has_value() && left_wpt &&
      HasLaneChange(left_marking->lane_change, LaneChange::Left) &&
      (wp_lane_id * left_wpt->GetLaneId() > 0) &&
      left_wpt->GetType() == road::Lane::LaneType::Driving) {
    auto check = VehicleObstacleDetected(vehicles_alist,
                                         behind_max,
                                         /*up_angle_th=*/180.0f,
                                         /*low_angle_th=*/0.0f,
                                         /*lane_offset=*/-1);
    if (!check.affected) {
      std::cout << "Tailgating, moving to the left!" << std::endl;
      auto end_waypoint = _local_planner->TargetWaypoint();
      _behavior.tailgate_counter = 200;
      if (end_waypoint) {
        SetDestination(end_waypoint->GetTransform().location,
                       left_wpt->GetTransform().location);
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Vehicle / pedestrian managers
// ---------------------------------------------------------------------------

BasicAgent::VehicleObstacle BehaviorAgent::CollisionAndCarAvoidManager(
    SharedPtr<client::Waypoint> waypoint) {
  auto vehicles_alist = _world->GetActors()
      ? _world->GetActors()->Filter("*vehicle*")
      : nullptr;

  // Python applies a 45 m prefilter around the waypoint here (excluding the
  // ego). The base helper `VehicleObstacleDetected` honors `max_distance`
  // internally, so we just route the world-wide vehicle list to it.
  VehicleObstacle obs;
  if (_direction == RoadOption::ChangeLaneLeft) {
    obs = VehicleObstacleDetected(
        vehicles_alist,
        std::max(_behavior.min_proximity_threshold, _speed_limit / 2.0f),
        /*up_angle_th=*/180.0f,
        /*low_angle_th=*/0.0f,
        /*lane_offset=*/-1);
  } else if (_direction == RoadOption::ChangeLaneRight) {
    obs = VehicleObstacleDetected(
        vehicles_alist,
        std::max(_behavior.min_proximity_threshold, _speed_limit / 2.0f),
        /*up_angle_th=*/180.0f,
        /*low_angle_th=*/0.0f,
        /*lane_offset=*/1);
  } else {
    obs = VehicleObstacleDetected(
        vehicles_alist,
        std::max(_behavior.min_proximity_threshold, _speed_limit / 3.0f),
        /*up_angle_th=*/30.0f,
        /*low_angle_th=*/0.0f,
        /*lane_offset=*/0);

    // Tailgating check.
    if (!obs.affected &&
        _direction == RoadOption::LaneFollow &&
        !waypoint->IsJunction() &&
        _speed > 10.0f &&
        _behavior.tailgate_counter == 0) {
      Tailgating(waypoint);
    }
  }

  return obs;
}

BasicAgent::VehicleObstacle BehaviorAgent::PedestrianAvoidManager(
    SharedPtr<client::Waypoint> waypoint) {
  auto walkers_alist = _world->GetActors()
      ? _world->GetActors()->Filter("*walker.pedestrian*")
      : nullptr;

  // Python applies a 10 m prefilter around the waypoint here. The base
  // `VehicleObstacleDetected` honors `max_distance` internally, so we let
  // it do the proximity gating from the world-wide walker list.
  (void)waypoint;  // location consumed indirectly via VehicleObstacleDetected
  VehicleObstacle obs;
  if (_direction == RoadOption::ChangeLaneLeft) {
    obs = VehicleObstacleDetected(
        walkers_alist,
        std::max(_behavior.min_proximity_threshold, _speed_limit / 2.0f),
        /*up_angle_th=*/90.0f,
        /*low_angle_th=*/0.0f,
        /*lane_offset=*/-1);
  } else if (_direction == RoadOption::ChangeLaneRight) {
    obs = VehicleObstacleDetected(
        walkers_alist,
        std::max(_behavior.min_proximity_threshold, _speed_limit / 2.0f),
        /*up_angle_th=*/90.0f,
        /*low_angle_th=*/0.0f,
        /*lane_offset=*/1);
  } else {
    obs = VehicleObstacleDetected(
        walkers_alist,
        std::max(_behavior.min_proximity_threshold, _speed_limit / 3.0f),
        /*up_angle_th=*/60.0f,
        /*low_angle_th=*/0.0f,
        /*lane_offset=*/0);
  }
  return obs;
}

// ---------------------------------------------------------------------------
// Car-following control law
// ---------------------------------------------------------------------------

rpc::VehicleControl BehaviorAgent::CarFollowingManager(
    SharedPtr<client::Actor> vehicle, float distance, bool debug) {
  const float vehicle_speed = GetSpeed(vehicle);
  const float delta_v = std::max(1.0f, (_speed - vehicle_speed) / 3.6f);
  const float ttc = (delta_v != 0.0f)
      ? distance / delta_v
      : distance / std::nextafter(0.0f, 1.0f);

  float target_speed = 0.0f;
  if (_behavior.safety_time > ttc && ttc > 0.0f) {
    target_speed = std::min({
        Positive(vehicle_speed - _behavior.speed_decrease),
        _behavior.max_speed,
        _speed_limit - _behavior.speed_lim_dist});
  } else if (2.0f * _behavior.safety_time > ttc && ttc >= _behavior.safety_time) {
    target_speed = std::min({
        std::max(_min_speed, vehicle_speed),
        _behavior.max_speed,
        _speed_limit - _behavior.speed_lim_dist});
  } else {
    target_speed = std::min({
        _behavior.max_speed,
        _speed_limit - _behavior.speed_lim_dist});
  }
  _local_planner->SetSpeed(target_speed);
  return _local_planner->RunStep(debug);
}

// ---------------------------------------------------------------------------
// Main run-step
// ---------------------------------------------------------------------------

rpc::VehicleControl BehaviorAgent::RunStep(bool debug) {
  UpdateInformation();

  if (_behavior.tailgate_counter > 0) {
    _behavior.tailgate_counter -= 1;
  }

  const auto ego_vehicle_loc = _vehicle->GetLocation();
  auto ego_vehicle_wp = _map->GetWaypoint(ego_vehicle_loc);

  // 1. Red lights / stops.
  if (TrafficLightManager()) {
    return EmergencyStop();
  }

  // 2.1: Pedestrian avoidance.
  auto walker_obs = PedestrianAvoidManager(ego_vehicle_wp);
  if (walker_obs.affected && walker_obs.actor) {
    const auto &walker_extent = walker_obs.actor->GetBoundingBox().extent;
    const auto &ego_extent    = _vehicle->GetBoundingBox().extent;
    const float distance = walker_obs.distance
        - std::max(walker_extent.y, walker_extent.x)
        - std::max(ego_extent.y,    ego_extent.x);
    if (distance < _behavior.braking_distance) {
      return EmergencyStop();
    }
  }

  // 2.2: Vehicle avoidance / car following.
  auto vehicle_obs = CollisionAndCarAvoidManager(ego_vehicle_wp);
  rpc::VehicleControl control{};
  if (vehicle_obs.affected && vehicle_obs.actor) {
    const auto &v_extent  = vehicle_obs.actor->GetBoundingBox().extent;
    const auto &ego_extent = _vehicle->GetBoundingBox().extent;
    const float distance = vehicle_obs.distance
        - std::max(v_extent.y,   v_extent.x)
        - std::max(ego_extent.y, ego_extent.x);
    if (distance < _behavior.braking_distance) {
      return EmergencyStop();
    }
    control = CarFollowingManager(vehicle_obs.actor, distance, debug);
  }
  // 3: Intersection behaviour (left/right turn).
  else if (_incoming_waypoint && _incoming_waypoint->IsJunction() &&
           (_incoming_direction == RoadOption::Left ||
            _incoming_direction == RoadOption::Right)) {
    const float target_speed = std::min(_behavior.max_speed,
                                        _speed_limit - 5.0f);
    _local_planner->SetSpeed(target_speed);
    control = _local_planner->RunStep(debug);
  }
  // 4: Normal behaviour.
  else {
    const float target_speed = std::min(_behavior.max_speed,
                                        _speed_limit - _behavior.speed_lim_dist);
    _local_planner->SetSpeed(target_speed);
    control = _local_planner->RunStep(debug);
  }
  return control;
}

// ---------------------------------------------------------------------------
// Emergency stop
// ---------------------------------------------------------------------------

rpc::VehicleControl BehaviorAgent::EmergencyStop() const {
  rpc::VehicleControl control{};
  control.throttle = 0.0f;
  control.brake = _max_brake;
  control.hand_brake = false;
  return control;
}

} // namespace navigation
} // namespace agents
} // namespace carla
